### PR TITLE
login function

### PR DESCRIPTION
**Pull request by trungdt705**
## What does this PR do?

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Jira ticket number?

Jira Ticket  (YOUR_TICKET_NUMBER_HERE)

## Checklist before merging

- [ ]
- [ ]

### CODE REVIEW

File: apps/frontend-web/app/login/page.tsx [hunk 1]
```diff
@@ -42,18 +50,43 @@ export default function Page() {
     mode: 'onChange',
   });
```
Comment (hiennq99) [inline apps/frontend-web/app/login/page.tsx:51]: separate to validation file

File: apps/frontend-web/app/login/page.tsx [hunk 2]
```diff
@@ -42,18 +50,43 @@ export default function Page() {
...
+          password: data.password,
+          provider: 'local',
+        },
+      });
+      if (result.login) {
+        storeUser({
+          token: result?.login.jwt,
+          user: {
+            id: result?.login.user?.id,
+            email: result?.login.user?.email,
```
Comment (hiennq99) [inline apps/frontend-web/app/login/page.tsx:67]: remove all nullable conditional. result.login.jwt,

File: apps/frontend-web/app/login/page.tsx [hunk 3]
```diff
@@ -42,18 +50,43 @@ export default function Page() {
...
+  const onSubmit = async (data: LoginForm) => {
+    try {
+      const result = await apiClient.login({
+        input: {
+          email: data.email,
+          password: data.password,
+          provider: 'local',
+        },
+      });
+      if (result.login) {
```
Comment (hiennq99) [inline apps/frontend-web/app/login/page.tsx:62]: if (result && result.login) or if (result?.login)

File: packages/ui-auth-protect/src/hooks/useAuth.tsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
+'use client';
+import { useState, useEffect } from 'react';
+import { LoginResult } from '../context/ui-auth-protect';
+
+export default function useAuthProvider() {
+  const [user, setUser] = useState<string | null>(null);
+  const [token, setToken] = useState<string | null>(null);
+  const [loading, setLoading] = useState<boolean>(false);
+
+  const storeUser = (data: LoginResult) => {
```
Comment (hiennq99) [inline packages/ui-auth-protect/src/hooks/useAuth.tsx:10]: rename to SetUser

File: apps/frontend-web/app/route-guard.tsx [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import { useAuth } from '@test-assessment/ui-auth-protect';
+import { useRouter } from 'next/navigation';
+import { useEffect } from 'react';
+
+export default function RouteGuard({ children }) {
+  const auth = useAuth();
+  const router = useRouter();
+
+  useEffect(() => {
+    if (auth.loading && !auth.user) {
```
Comment (hiennq99) [inline apps/frontend-web/app/route-guard.tsx:11]: why do you need to check auth.loading?

File: apps/frontend-web/app/route-guard.tsx [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+      router.push('/login');
+    }
+  }, [auth]);
```
Comment (hiennq99) [inline apps/frontend-web/app/route-guard.tsx:14]: only check auth.loading or auth.user. probably auth will have many properties can be changed

File: package.json [hunk 7]
```diff
@@ -14,6 +14,7 @@
     "next": "13.1.1",
     "react": "18.2.0",
     "react-dom": "18.2.0",
+    "styled-jsx": "5.1.2",
```
Comment (hiennq99) [inline package.json:17]: we are not using this library

File: packages/ui-auth-protect/src/hooks/useAuth.tsx [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    if (window && typeof window !== 'undefined') {
+      localStorage.setItem('user', JSON.stringify(data?.user));
+      localStorage.setItem('token', data?.token);
+    }
+  };
+  useEffect(() => {
+    async function loadUser() {
+      // simulate get user info
+      // TODO: get user info to verify token
+      if (window && typeof window !== 'undefined') {
```
Comment (hiennq99) [inline packages/ui-auth-protect/src/hooks/useAuth.tsx:22]: why do you check double check value of window?

**Review by hiennq99 — CHANGES_REQUESTED** [review]

File: apps/frontend-web/app/login/page.tsx (hunk 1 above)
Comment (trungdt705) [reply apps/frontend-web/app/login/page.tsx:37]: resolved, using validation.tsx file in login folder

File: apps/frontend-web/app/login/page.tsx (hunk 2 above)
Comment (trungdt705) [reply apps/frontend-web/app/login/page.tsx]: resolved

File: apps/frontend-web/app/login/page.tsx (hunk 3 above)
Comment (trungdt705) [reply apps/frontend-web/app/login/page.tsx]: resolved

File: packages/ui-auth-protect/src/hooks/useAuth.tsx (hunk 4 above)
Comment (trungdt705) [reply packages/ui-auth-protect/src/hooks/useAuth.tsx]: resolved

File: apps/frontend-web/app/route-guard.tsx (hunk 5 above)
Comment (trungdt705) [reply apps/frontend-web/app/route-guard.tsx]: Maybe, we don't need auth.loading. removed

File: apps/frontend-web/app/route-guard.tsx (hunk 6 above)
Comment (trungdt705) [reply apps/frontend-web/app/route-guard.tsx:15]: yep, removed auth.loading

File: package.json (hunk 7 above)
Comment (trungdt705) [reply package.json]: removed. I don't know why this dependency appear. so weird

File: packages/ui-auth-protect/src/hooks/useAuth.tsx (hunk 8 above)
Comment (trungdt705) [reply packages/ui-auth-protect/src/hooks/useAuth.tsx]: yep, we just need check typeof window. thanks

**Review by chungnq253 — CHANGES_REQUESTED** [review]

File: packages/ui-auth-protect/src/context/ui-auth-protect.tsx [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+  token: string;
+  user: {
+    id: string;
+    email: string;
+  };
+};
+
+const authContext = createContext({
+  user: typeof window !== 'undefined' ? localStorage.getItem('user') : null,
+  SetUser: (data: LoginResult) => {},
```
[the remark this answers is not in the document]
Comment (trungdt705) [reply packages/ui-auth-protect/src/context/ui-auth-protect.tsx:16]: duplicated with  const [user, setUser] = useState<string | null>(
    localStorage.getItem('user') || null
  ); in my code. Then, i used storeUser function but Hien thought that we need change to SetUser, I will revert this function to storeUser. Let me know if you have any good function name. Thanks

File: packages/ui-auth-protect/src/hooks/useAuth.tsx [hunk 10]
```diff
@@ -0,0 +1,36 @@
+'use client';
+import { useState, useEffect } from 'react';
+import { LoginResult } from '../context/ui-auth-protect';
+
+export default function useAuthProvider() {
+  const [user, setUser] = useState<string | null>(
+    typeof window !== 'undefined' ? localStorage.getItem('user') : null
```
[the remark this answers is not in the document]
Comment (trungdt705) [reply packages/ui-auth-protect/src/hooks/useAuth.tsx:7]: i see an error an terminal log like this "ReferenceError: localStorage is not defined"

File: packages/ui-auth-protect/src/context/ui-auth-protect.tsx [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+export type LoginResult = {
+  token: string;
+  user: {
+    id: string;
+    email: string;
+  };
+};
+
+const authContext = createContext({
+  user: typeof window !== 'undefined' ? localStorage.getItem('user') : null,
```
[the remark this answers is not in the document]
Comment (trungdt705) [reply packages/ui-auth-protect/src/context/ui-auth-protect.tsx:15]: i see an error an terminal log like this "ReferenceError: localStorage is not defined"

File: packages/cms-graphql-api/src/provider/index.tsx [hunk 12]
```diff
@@ -23,7 +23,7 @@ const getHeaders = (token?: string | null) => {
 
 export const getApiClient = (
   // eslint-disable-next-line turbo/no-undeclared-env-vars
-  apiUrl = process.env.GRAPHQL_API_URL || '/graphql',
+  apiUrl = 'http://localhost:1337/graphql',
```
[the remark this answers is not in the document]
Comment (trungdt705) [reply packages/cms-graphql-api/src/provider/index.tsx:26]: i will remove it. thanks

File: apps/frontend-web/app/route-guard.tsx [hunk 13]
```diff
@@ -0,0 +1,17 @@
+'use client';
+import { useAuth } from '@test-assessment/ui-auth-protect';
+import { useRouter } from 'next/navigation';
+import { useEffect } from 'react';
+
+export default function RouteGuard({ children }) {
+  const auth = useAuth();
+  const router = useRouter();
+  console.log(auth);
```
[the remark this answers is not in the document]
Comment (trungdt705) [reply apps/frontend-web/app/route-guard.tsx:9]: ok. thanks

File: apps/frontend-web/app/login/page.tsx [hunk 14]
```diff
@@ -42,18 +36,43 @@ export default function Page() {
...
+    try {
+      const result = await apiClient.login({
+        input: {
+          email: data.email,
+          password: data.password,
+          provider: 'local',
+        },
+      });
+      if (result?.login) {
+        SetUser({
```
[the remark this answers is not in the document]
Comment (trungdt705) [reply apps/frontend-web/app/login/page.tsx:49]: duplicated with const [user, setUser] = useState<string | null>(
localStorage.getItem('user') || null
); in my code. Then, i used storeUser function but Hien thought that we need change to SetUser, I will revert this function to storeUser. Let me know if you have any good function name. Thanks

File: yarn.lock [hunk 15]
```diff
@@ -21185,6 +21185,13 @@ styled-jsx@5.1.1:
   dependencies:
     client-only "0.0.1"
 
+styled-jsx@5.1.2:
```
[the remark this answers is not in the document]
Comment (trungdt705) [reply yarn.lock:21188]: sure. thanks

**Review by hiennq99 — APPROVED** [review]

**Review by chungnq253 — APPROVED** [review]

**Pull request merged**